src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java

```diff
@@ -0,0 +1,199 @@
+            if (c == null) {
+                return null;
+            }
+            synchronized (secrets) {
+                if (secrets.containsKey(c)) {
+                    Set<String> values = secrets.get(c);
+                    if (values != null) {
+                        LOGGER.log(Level.FINE, "Using cached secrets for {0}", c);
+                        return new SecretsMasker(values);
+                    } else {
+                        LOGGER.log(Level.FINE, "Cached absence of secrets for {0}", c);
+                        return null;
+                    }
+                }
+            }
+            Set<String> values = secretsOf(c);
+            synchronized (secrets) {
+                secrets.put(c, values);
+            }
+            if (values != null) {
+                LOGGER.fine("masking " + values.size() + " values");
+                return TaskListenerDecorator.merge(context.get(TaskListenerDecorator.class), new SecretsMasker(values));
+            } else {
+                LOGGER.fine("no values to mask");
+                return null;
+            }
+        }
+
+        private static @CheckForNull Set<String> secretsOf(KubernetesComputer c) throws IOException, InterruptedException {
+            Channel ch = c.getChannel();
+            if (ch == null) {
+                return null;
+            }
+            KubernetesSlave slave = c.getNode();
+            if (slave == null) {
+                return null;
+            }
+            Pod pod = slave.getTemplate().build(slave);
+            Set<String> values = new HashSet<>();
+            LOGGER.finer(() -> "inspecting " + pod);
+            for (Container container : pod.getSpec().getContainers()) {
+                Set<String> secretContainerKeys = new TreeSet<>();
+                for (EnvVar envVar : container.getEnv()) {
+                    EnvVarSource envVarSource = envVar.getValueFrom();
+                    if (envVarSource != null) {
+                        SecretKeySelector secretKeySelector = envVarSource.getSecretKeyRef();
+                        if (secretKeySelector != null) {
+                            secretContainerKeys.add(envVar.getName());
+                        }
+                    }
+                }
+                if (!secretContainerKeys.isEmpty()) {
+                    String containerName = container.getName();
+                    LOGGER.fine(() -> "looking for " + slave.getNamespace() + "/" + slave.getPodName() + "/" + containerName + " secrets named " + secretContainerKeys);
+                    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                    Semaphore semaphore = new Semaphore(0);
+                    try (ExecWatch exec = slave.getKubernetesCloud().connect().pods().inNamespace(slave.getNamespace()).withName(slave.getPodName()).inContainer(containerName)
+                            .writingOutput(baos).writingError(System.err).writingErrorChannel(System.err)
```

Comment:
`System.err` leftover from early attempts?

Reply:
Not a leftover, but could be improved.